ErrorBoundary subclassing and context (#33585)
dotnet/AspNetCore.Docs: aspnetcore/blazor/fundamentals/handle-errors.md

For the preceding example, the app's stylesheet presumably includes an `errorUI` CSS class to style the content. The error content is rendered from the <xref:Microsoft.AspNetCore.Components.ErrorBoundaryBase.ErrorContent> property without a block-level element. A block-level element, such as a division (`<div>`) or a paragraph (`<p>`) element, can wrap the error content markup, but it isn't required.

Optionally, use the context (`@context`) of the <xref:Microsoft.AspNetCore.Components.ErrorBoundaryBase.ErrorContent> to obtain error data:

```razor
<ErrorContent>

>     @context.Message

Author: ```suggestion
    @context.HelpLink
```

</ErrorContent>
```

The <xref:Microsoft.AspNetCore.Components.ErrorBoundaryBase.ErrorContent> can also name the context. In the following example, the context is named `exception`:

```razor
<ErrorContent Context="exception">

>     @exception.Message

Author: ```suggestion
    @exception.HelpLink
```

* The user performs a UI gesture, such as selecting a button to indicate that they want to retry a procedure or when the user navigates to a new component.
* Additional logic that executes also clears the exception. When the component is rerendered, the error doesn't reoccur.

The following example permits the user to recover from the exception with a button:

```razor
<ErrorBoundary @ref="errorBoundary">
    <ChildContent>
        <EmbeddedCounter />
    </ChildContent>
    <ErrorContent>
        <div class="alert alert-danger" role="alert">
            <p class="fs-3 fw-bold">😈 A rotten gremlin got us. Sorry!</p>

>             <p>@context.GetType(): @context.Message</p>

Author: Yes ... I thought about that but didn't state my usual warning about it. I guess I'm looking for ***anything*** that we can do there with the context to demo it. What can it show that would be safe? How about I change it to the `HelpLink`?

>             <p>@context.GetType(): @context.Message</p>

Author: ```suggestion
            <p>@context.HelpLink</p>
```

            <button class="btn btn-info" @onclick="_ => errorBoundary.Recover()">
                Clear
            </button>
        </div>
    </ErrorContent>
</ErrorBoundary>

@code {
    private ErrorBoundary? errorBoundary;
}
```

You can also subclass <xref:Microsoft.AspNetCore.Components.Web.ErrorBoundary> for custom processing by overriding <xref:Microsoft.AspNetCore.Components.Web.ErrorBoundary.OnErrorAsync%2A>. The following example merely logs the error, but you can implement any error handling code you wish. You can remove the line that awaits <xref:System.Threading.Tasks.Task.Yield%2A?displayProperty=nameWithType> if your code awaits some other asynchronous task.

`CustomErrorBoundary.razor`:

```razor
@inherits ErrorBoundary

> @inject ILogger<CustomErrorBoundary>? Logger

Author: I checked the whole Blazor node for this mistake ... this is the only instance of me goofing it 🙈👍.

@if (CurrentException is null)
{
    @ChildContent
}
else if (ErrorContent is not null)
{
    @ErrorContent(CurrentException)
}

@code {
    protected override async Task OnErrorAsync(Exception ex)
    {
        await Task.Yield();
        Logger?.LogError("{Type}: {Message}", ex.GetType(), ex.Message);
    }
}
```

The preceding example can also be implemented as a class.

`CustomErrorBoundary.cs`:

```csharp
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Web;

namespace BlazorSample;

public class CustomErrorBoundary : ErrorBoundary
{
    [Inject]
    ILogger<CustomErrorBoundary>? Logger {  get; set; }

    protected override async Task OnErrorAsync(Exception ex)
    {

>         await Task.Yield();
>         Logger?.LogError("{Type}: {Message}", ex.GetType(), ex.Message);

Author: ```suggestion
        Logger.LogError(ex, "An error occurred while rendering a component.");
        return Task.CompletedTask;
```
